server/coordinator/procedure/transfer_leader.go

```diff
@@ -0,0 +1,317 @@
+	request := event.Args[0].(*TransferLeaderCallbackRequest)
+	ctx := request.ctx
+
+	if request.cluster.GetClusterState() != storage.ClusterStateStable {
+		cancelEventWithLog(event, cluster.ErrClusterStateInvalid, "cluster state must be stable", zap.Int("currentState", int(request.cluster.GetClusterState())))
+		return
+	}
+
+	shardNodes, err := request.cluster.GetShardNodesByShardID(request.shardID)
+	if err != nil {
+		cancelEventWithLog(event, err, "get shardNodes by shardID failed")
+		return
+	}
+
+	found := false
+	var leaderShardNode storage.ShardNode
+	for _, shardNode := range shardNodes {
+		if shardNode.ShardRole == storage.ShardRoleLeader {
+			found = true
+			leaderShardNode = shardNode
+			leaderShardNode.NodeName = request.newLeaderNodeName
+		}
+	}
+	if !found {
+		cancelEventWithLog(event, ErrShardLeaderNotFound, "shard leader not found", zap.Uint32("shardID", uint32(request.shardID)))
+		return
+	}
+
+	err = request.cluster.UpdateClusterView(ctx, storage.ClusterStateStable, []storage.ShardNode{leaderShardNode})
+	if err != nil {
+		cancelEventWithLog(event, storage.ErrUpdateClusterViewConflict, "update cluster view failed")
+		return
+	}
+}
+
+func transferLeaderCloseOldLeaderCallback(event *fsm.Event) {
+	request := event.Args[0].(*TransferLeaderCallbackRequest)
+	ctx := request.ctx
+
+	closeShardRequest := eventdispatch.CloseShardRequest{
+		ShardID: uint32(request.shardID),
+	}
+	if err := request.dispatch.CloseShard(ctx, request.oldLeaderNodeName, closeShardRequest); err != nil {
+		cancelEventWithLog(event, err, "close shard failed", zap.Uint32("shardId", uint32(request.shardID)), zap.String("oldLeaderName", request.oldLeaderNodeName))
+		return
+	}
+}
+
+func transferLeaderOpenNewShardCallback(event *fsm.Event) {
+	request := event.Args[0].(*TransferLeaderCallbackRequest)
+	ctx := request.ctx
+
+	// TODO: Shard version is only exists in ShardTables now, it is not sensible. Because shard version will be update when shard node mapping changed or shard table mapping changed, so ShardNodes also need shard version, consider refactor later.
+	shardTablesMapping := request.cluster.GetTables([]storage.ShardID{request.shardID}, request.oldLeaderNodeName)
+	shardTables, exists := shardTablesMapping[request.shardID]
+	if !exists {
+		cancelEventWithLog(event, cluster.ErrTableNotFound, "tables on shard not found")
+		return
+	}
+
+	openShardRequest := eventdispatch.OpenShardRequest{
+		Shard: cluster.ShardInfo{ID: request.shardID, Role: storage.ShardRoleLeader, Version: shardTables.Shard.Version + 1},
+	}
+
+	if err := request.dispatch.OpenShard(ctx, request.newLeaderNodeName, openShardRequest); err != nil {
+		cancelEventWithLog(event, err, "open shard failed", zap.Uint32("shardId", uint32(request.shardID)), zap.String("newLeaderNode", request.newLeaderNodeName))
+		return
+	}
+}
+
+func transferLeaderFinishCallback(event *fsm.Event) {
+	request := event.Args[0].(*TransferLeaderCallbackRequest)
```

Comment:
Need to check index overflow and type assert.

Reply:
I think I don't think it is necessary to detect parameter types. If `event.Args[0]` is not `TransferLeaderCallbackRequest`, it will be fast fail in unit test.